src/profile-logic/symbol-store.js

```diff
@@ -350,57 +355,104 @@ export class SymbolStore {
+          }
 
-            // Store the symbol table in the database.
-            await this._storeSymbolTableInDB(lib, symbolTable);
-          } catch (error2) {
-            // None of the symbolication methods were successful.
-            // Call the error callback.
-            errorCb(
-              request,
-              new SymbolsNotFoundError(
-                `Could not obtain symbols for ${lib.debugName}/${lib.breakpadId}.`,
-                lib,
-                error1,
-                error2
-              )
+          if (requestsWithErrors.length > 0) {
+            // Fall back to getting symbols from the browser for these.
+            await this._getSymbolsFromBrowser(
+              requestsWithErrors,
+              demangleCallback,
+              successCb,
+              errorCb
             );
           }
+        } catch (error) {
+          // There was a problem for the entire chunk of requests, for example there
+          // might have been a problem parsing the response JSON.
+          // Map all requests to the same error, and then try to get symbols from the browser.
+          const requestsWithErrors = requests.map((request) => ({
+            request,
+            error: new SymbolsNotFoundError(
+              'There was a problem with the JSON returned by the symbolication API.',
+              request.lib,
+              error
+            ),
+          }));
+          await this._getSymbolsFromBrowser(
+            requestsWithErrors,
+            demangleCallback,
+            successCb,
+            errorCb
+          );
         }
       })
     );
   }
+
+  // Try to get symbols from the browser for any libraries which couldn't be
+  // symbolicated with the symbolication API.
+  async _getSymbolsFromBrowser(
+    requestsWithErrors: Array<{|
+      // A symbolication request for one library.
+      request: LibSymbolicationRequest,
+      // The error which was encountered when trying to obtain symbols for
+      // this library using the symbolication API.
+      // This gets reported in the final error if all avenues fail.
+      error: Error,
+    |}>,
+    demangleCallback: DemangleFunction,
+    successCb: (RequestedLib, Map<number, AddressResult>) => void,
+    errorCb: (LibSymbolicationRequest, Error) => void
+  ): Promise<void> {
+    for (const { request, error } of requestsWithErrors) {
+      const { lib, addresses } = request;
+      try {
+        // Option 3: Request a symbol table from the browser.
+        // This call will throw if the browser cannot obtain the symbol table.
+        const symbolTable =
+          await this._symbolProvider.requestSymbolTableFromBrowser(lib);
+
+        // Did not throw, option 3 was successful!
+        successCb(
+          lib,
+          readSymbolsFromSymbolTable(addresses, symbolTable, demangleCallback)
+        );
+
+        // Store the symbol table in the database.
+        await this._storeSymbolTableInDB(lib, symbolTable);
```

Comment:
BTW, I realize we don't store the symbol table in the DB in the "request from server" case, is that normal?
(out of scope for this PR)

Reply:
This is normal - we only store full symbol tables. When we request symbols from the server, we only get symbols for a subset of addresses, not for the entire library. So we'd need to have a new storage format with partial symbols, but I'm not sure if that's worth it. Furthermore, the symbol tables we store only have outer functions and no filename / line / inline information. So for any libraries that we get from the DB, we won't have that rich data.